jigglypuffRL/deeprl/ppo1/ppo1.py

```diff
@@ -0,0 +1,229 @@
+            .mul(A)
+            .to(self.device)
+        )
+
+        loss_policy = (
+            torch.mean(torch.min(torch.mul(ratio, A), clipping)).mul(-1).unsqueeze(0)
+        )
+        loss_value = nn.MSELoss()(
+            self.value_fn.value_hist, Variable(returns)
+        ).unsqueeze(0)
+
+        # store traj loss values in epoch loss tensors
+        self.policy_new.loss_hist = torch.cat([self.policy_new.loss_hist, loss_policy])
+        self.value_fn.loss_hist = torch.cat([self.value_fn.loss_hist, loss_value])
+
+        # clear traj history
+        self.policy_old.traj_reward = []
+        self.policy_old.policy_hist = Variable(torch.Tensor())
+        self.policy_new.policy_hist = Variable(torch.Tensor())
+        self.value_fn.value_hist = Variable(torch.Tensor())
+
+    def update_policy(self, ep):
+        # mean of all traj losses in single epoch
+        loss_policy = torch.mean(self.policy_new.loss_hist)
+        loss_value = torch.mean(self.value_fn.loss_hist)
+
+        # tensorboard book-keeping
+        if self.tensorboard_log:
+            self.writer.add_scalar("loss/policy", loss_policy, ep)
+            self.writer.add_scalar("loss/value", loss_value, ep)
+
+        # take gradient step
+        self.optimizer_policy.zero_grad()
+        loss_policy.backward()
+        self.optimizer_policy.step()
+
+        self.optimizer_value.zero_grad()
+        loss_value.backward()
+        self.optimizer_value.step()
+
+        # clear loss history for epoch
+        self.policy_new.loss_hist = Variable(torch.Tensor())
+        self.value_fn.loss_hist = Variable(torch.Tensor())
+
+    def learn(self):
+        # training loop
+        for ep in range(self.epochs):
+            epoch_reward = 0
+            for i in range(self.actor_batch_size):
+                s = self.env.reset()
+                done = False
+                for t in range(self.timesteps_per_actorbatch):
+                    a = self.select_action(s)
+                    s, r, done, _ = self.env.step(np.array(a))
+
+                    self.policy_old.traj_reward.append(r)
+
+                    if done:
+                        break
+
+                epoch_reward += (
+                    np.sum(self.policy_old.traj_reward) / self.actor_batch_size
+                )
+                self.get_traj_loss()
+
+            self.update_policy(ep)
+
+            if ep % 20 == 0:
+                print("Episode: {}, reward: {}".format(ep, epoch_reward))
+                if self.tensorboard_log:
+                    self.writer.add_scalar("reward", epoch_reward, ep)
+
+            if ep % self.policy_copy_interval == 0:
+                self.policy_old.load_state_dict(self.policy_new.state_dict())
+
+        if self.tensorboard_log:
+            self.writer.close()
+
+
+if __name__ == "__main__":
+    env = gym.make("LunarLander-v2")
+    algo = PPO1("MlpPolicy", "MlpValue", env)
+    algo.learn()
```

Comment:
We would probably need a command line parser here, but maybe we can add that later, thats ok for now. For specifying options to train/test, render, etc.

Reply:
Yeah good point. this is just for other's to refer usage while working on other algorithms initially. Will be removed later

Reply:
Will resolve this for now. Adding as an issue